lib/Frontend/Frontend.cpp

```diff
@@ -138,70 +158,122 @@ bool CompilerInstance::setup(const CompilerInvocation &Invok) {
-    CodeCompletionBufferID = SourceMgr.addMemBufferCopy(MemBuf);
-    InputSourceCodeBufferIDs.push_back(*CodeCompletionBufferID);
-    SourceMgr.setCodeCompletionPoint(*CodeCompletionBufferID,
-                                     CodeCompletePoint.second);
+    codeCompletionBufferID = SourceMgr.addMemBufferCopy(memBuf);
+    InputSourceCodeBufferIDs.push_back(*codeCompletionBufferID);
+    SourceMgr.setCodeCompletionPoint(*codeCompletionBufferID,
+                                     codeCompletePoint.second);
   }
+  return codeCompletionBufferID;
+}
 
-  bool MainMode = (Invocation.getInputKind() == InputFileKind::IFK_Swift);
-  bool SILMode = (Invocation.getInputKind() == InputFileKind::IFK_SIL);
+bool CompilerInstance::setupInputs(Optional<unsigned> codeCompletionBufferID) {
+  // Add the memory buffers first, these will be associated with a filename
+  // and they can replace the contents of an input filename.
+  for (const InputFile &input :
+       Invocation.getFrontendOptions().Inputs.getAllFiles())
+    if (setupForInput(input))
+      return true;
 
-  if (SILMode)
-    Invocation.getLangOptions().EnableAccessControl = false;
+  // Set the primary file to the code-completion point if one exists.
+  if (codeCompletionBufferID.hasValue())
+    PrimaryBufferID = *codeCompletionBufferID;
 
-  const Optional<SelectedInput> &PrimaryInput =
-      Invocation.getFrontendOptions().Inputs.getOptionalPrimaryInput();
+  if (isInMainMode() && MainBufferID == NO_SUCH_BUFFER &&
+      InputSourceCodeBufferIDs.size() == 1)
+    MainBufferID = InputSourceCodeBufferIDs.front();
 
-  // Add the memory buffers first, these will be associated with a filename
-  // and they can replace the contents of an input filename.
-  for (unsigned i = 0,
-                e = Invocation.getFrontendOptions().Inputs.inputBufferCount();
-       i != e; ++i) {
-    // CompilerInvocation doesn't own the buffers, copy to a new buffer.
-    auto *InputBuffer =
-        Invocation.getFrontendOptions().Inputs.getInputBuffers()[i];
-    auto Copy = std::unique_ptr<llvm::MemoryBuffer>(
-        llvm::MemoryBuffer::getMemBufferCopy(
-            InputBuffer->getBuffer(), InputBuffer->getBufferIdentifier()));
-    if (serialization::isSerializedAST(Copy->getBuffer())) {
-      PartialModules.push_back({ std::move(Copy), nullptr });
-    } else {
-      unsigned BufferID = SourceMgr.addNewSourceBuffer(std::move(Copy));
-      InputSourceCodeBufferIDs.push_back(BufferID);
+  return false;
+}
+
+bool CompilerInstance::setupForInput(const InputFile &input) {
+  if (llvm::MemoryBuffer *inputBuffer = input.getBuffer()) {
+    setupForBuffer(inputBuffer, input.getIsPrimary());
+    return false;
+  }
+  return setUpForFile(input.getFile(), input.getIsPrimary());
+}
+void CompilerInstance::setupForBuffer(llvm::MemoryBuffer *buffer,
+                                      bool isPrimary) {
+  auto copy =
+      std::unique_ptr<llvm::MemoryBuffer>(llvm::MemoryBuffer::getMemBufferCopy(
```

Comment:
This is already the return type of `getMemBufferCopy` these days (it used to be a bare `llvm::MemoryBuffer *`). May as well clean that up.

Reply:
Fixed.